File: web/src/lib/geo/haversine.ts
/**
 * Great-circle distance between two WGS84 coordinates in metres.
 *
 * We deliberately use the Haversine formula (not Vincenty / geodetic) - the
 * customer pin and the agent's GPS fix are both already noisy at the metre
 * level, and Haversine is good to ~0.5% which is at least an order of
 * magnitude tighter than the noise we care about. Stays simple and fast.
 *
 * R = 6_371_000 m is the WGS84 mean Earth radius.
 */
export function haversineMeters(
	a: { lat: number; lng: number },
	b: { lat: number; lng: number },
): number {
	const R = 6_371_000;
	const toRad = (deg: number) => (deg * Math.PI) / 180;

	const dLat = toRad(b.lat - a.lat);
	const dLng = toRad(b.lng - a.lng);
	const lat1 = toRad(a.lat);
	const lat2 = toRad(b.lat);

	const sinDLat = Math.sin(dLat / 2);
	const sinDLng = Math.sin(dLng / 2);

	const h =
		sinDLat * sinDLat + Math.cos(lat1) * Math.cos(lat2) * sinDLng * sinDLng;
	return 2 * R * Math.asin(Math.min(1, Math.sqrt(h)));
}

File: android/app/src/main/java/com/threefat/vcts/domain/geo/Haversine.kt
package com.threefat.vcts.domain.geo

import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Great-circle distance in metres. Bit-for-bit identical to the server's
 * `web/src/lib/geo/haversine.ts` so the client's pre-flight check matches
 * the authoritative server validation - if we say "you're inside the
 * fence" and unlock the form, the POST will not fail with a geofence
 * violation.
 *
 * `R = 6_371_000 m` is the WGS84 mean Earth radius. Accuracy is ~0.5%, an
 * order of magnitude better than typical GPS noise.
 */
fun haversineMeters(
    aLat: Double,
    aLng: Double,
    bLat: Double,
    bLng: Double,
): Double {
    val r = 6_371_000.0

    fun rad(deg: Double) = deg * Math.PI / 180.0

    val dLat = rad(bLat - aLat)
    val dLng = rad(bLng - aLng)
    val lat1 = rad(aLat)
    val lat2 = rad(bLat)

    val sinDLat = sin(dLat / 2)
    val sinDLng = sin(dLng / 2)

    val h = sinDLat * sinDLat + cos(lat1) * cos(lat2) * sinDLng * sinDLng
    return 2 * r * asin(min(1.0, sqrt(h)))
}
